crates/anvil/src/eth/backend/mem/state.rs

```diff
@@ -1,81 +1,66 @@
-use alloy_primitives::{Address, Bytes, B256, U256};
+use crate::eth::error::BlockchainError;
+use alloy_primitives::{keccak256, Address, Bytes, B256, U256};
 use alloy_rlp::Encodable;
 use alloy_rpc_types::state::StateOverride;
-use anvil_core::eth::trie::RefSecTrieDBMut;
+use alloy_trie::{HashBuilder, Nibbles};
 use foundry_evm::{
     backend::DatabaseError,
     revm::{
         db::{CacheDB, DatabaseRef, DbAccount},
         primitives::{AccountInfo, Bytecode, HashMap},
     },
 };
-use memory_db::HashKey;
-use trie_db::TrieMut;
+use itertools::Itertools;
 
-/// Returns storage trie of an account as `HashDB`
-pub fn storage_trie_db(storage: &HashMap<U256, U256>) -> (AsHashDB, B256) {
-    // Populate DB with full trie from entries.
-    let (db, root) = {
-        let mut db = <memory_db::MemoryDB<_, HashKey<_>, _>>::default();
-        let mut root = Default::default();
-        {
-            let mut trie = RefSecTrieDBMut::new(&mut db, &mut root);
-            for (k, v) in storage.iter().filter(|(_k, v)| *v != &U256::from(0)) {
-                let key = B256::from(*k);
-                let mut value: Vec<u8> = Vec::new();
-                U256::encode(v, &mut value);
-                trie.insert(key.as_slice(), value.as_ref()).unwrap();
-            }
-        }
-        (db, root)
-    };
-
-    (Box::new(db), B256::from(root))
+pub fn build_root(values: impl IntoIterator<Item = (Nibbles, Bytes)>) -> B256 {
+    let mut builder = HashBuilder::default();
+    for (key, value) in values {
+        builder.add_leaf(key, value.as_ref());
+    }
+    builder.root()
 }
 
-/// Returns the account data as `HashDB`
-pub fn trie_hash_db(accounts: &HashMap<Address, DbAccount>) -> (AsHashDB, B256) {
-    let accounts = trie_accounts(accounts);
+/// Builds state root from the given accounts
+pub fn state_root(accounts: &HashMap<Address, DbAccount>) -> B256 {
+    build_root(trie_accounts(accounts))
+}
 
-    // Populate DB with full trie from entries.
-    let (db, root) = {
-        let mut db = <memory_db::MemoryDB<_, HashKey<_>, _>>::default();
-        let mut root = Default::default();
-        {
-            let mut trie = RefSecTrieDBMut::new(&mut db, &mut root);
-            for (address, value) in accounts {
-                trie.insert(address.as_ref(), value.as_ref()).unwrap();
-            }
-        }
-        (db, root)
-    };
+/// Builds storage root from the given storage
+pub fn storage_root(storage: &HashMap<U256, U256>) -> B256 {
+    build_root(trie_storage(storage))
+}
 
-    (Box::new(db), B256::from(root))
+/// Builds iterator over stored key-value pairs ready for storage trie root calculation.
+pub fn trie_storage(storage: &HashMap<U256, U256>) -> Vec<(Nibbles, Bytes)> {
+    storage
+        .iter()
+        .map(|(key, value)| {
+            let data = alloy_rlp::encode(value);
+            (Nibbles::unpack(keccak256(key.to_be_bytes_vec())), data.into())
+        })
+        .sorted_by_key(|(key, _)| key.clone())
```

Comment:
Should be `x = iter map collect; x.sort_by(a.cmp(b)); x`
1. itertools `sorted` collects then does `into_iter` again
2. use `sort_by` to not clone, this is a current limitation of the `by_key` set of functions which I don't know if it will ever be fixed